### PR TITLE
Fix #10439 - Problems building with 3.12 on GHA - Bump python to 3.12.2 

### LINKED ISSUE #10439: Problems building with 3.12 on GHA
Issue overview
--------------

The macOS packages build fine, but EPTravisTester (also confirmed locally on my own M1) refuse to build due to "invalid code signature" on the python dylib.

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version): macOS
 - Version of EnergyPlus (if using an intermediate build, include SHA): develop, with 3.12
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------
 - Fixes #10439
- Supersedes #10432
- Remove macos-11 runner, will be deleted from github on June 28 (in 18 days)
- add Ubuntu 24.04 x64 runner
- add macos-14 (arm64) runner
- Bump to python 3.12.2

I forked actions/setup-python to https://github.com/jmarrec/setup-python which uses my forked https://github.com/jmarrec/python-versions/ that, amongst other things, force builds python on macOS, which solves the issue.
I am trying to upstream some of these changes to the official actions/python-versions but cannot guarantee they will be approved. I have opened 3 PRs so far: https://github.com/actions/python-versions/pulls/jmarrec

I have prebuilt 3.12.2 (and 3.8.10), but could build others as needed.

I have made a metric ton of test releases, but the last one that is ok (aside from the 24.04 _test_ step, but that's just an easy fix that I pushed after) is at https://github.com/jmarrec/EnergyPlus/releases/tag/Test-Python312-2024-06-10

**Edit**:  With 24.04 test step fixed: https://github.com/jmarrec/EnergyPlus/releases/tag/Test-Python312-2024-06-10-2

Run: 

* Linux: https://github.com/jmarrec/EnergyPlus/actions/runs/9448142447
* Mac: https://github.com/jmarrec/EnergyPlus/actions/runs/9448142454
* Windows: https://github.com/jmarrec/EnergyPlus/actions/runs/9448142449

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by jmarrec** [comment]
@Myoldmopar decent ci does not explicitly pass `-DPython_REQUIRED_VERSION`, so it defaults to looking for 3.12.2, which fails because the runner doesn't have it.

Either pass `-DPython_REQUIRED_VERSION:STRING=3.8` (or something) or better yet, make sure 3.12.2 is installed on the runners.

File: CMakeLists.txt [hunk 1]
```diff
@@ -191,7 +191,7 @@ if(LINK_WITH_PYTHON)
   if(Python_REQUIRED_VERSION)
     find_package(Python ${Python_REQUIRED_VERSION} EXACT COMPONENTS Interpreter Development REQUIRED)
   else()
-    find_package(Python 3.8 COMPONENTS Interpreter Development REQUIRED)
+    find_package(Python 3.12 COMPONENTS Interpreter Development REQUIRED)
```
Comment (jmarrec) [inline CMakeLists.txt:194]: This makes decent ci fail.

File: scripts/package_tests/runner.py [hunk 2]
```diff
@@ -86,15 +86,15 @@ class OS:
     'ubuntu2204': {
         'os': OS.Linux, 'bitness': 'x64', 'asset_pattern': 'Linux-Ubuntu22.04-x86_64.tar.gz', 'os_version': '22.04'
     },
+    'ubuntu2404': {
+        'os': OS.Linux, 'bitness': 'x64', 'asset_pattern': 'Linux-Ubuntu24.04-x86_64.tar.gz', 'os_version': '24.04'
+    },
```
Comment (jmarrec) [inline scripts/package_tests/runner.py:91]: Add ubuntu24.04

File: .github/workflows/linux_release.yml [hunk 3]
```diff
@@ -8,7 +8,7 @@ on:
 env:
   GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
   BUILD_TYPE: Release
-  Python_REQUIRED_VERSION: 3.8
+  Python_REQUIRED_VERSION: 3.12.2
```
Comment (jmarrec) [inline .github/workflows/linux_release.yml:11]: Bump

File: .github/workflows/linux_release.yml [hunk 4]
```diff
@@ -30,7 +30,7 @@ jobs:
 
     - name: Set up Python ${{ env.Python_REQUIRED_VERSION }}
       id: setup-python
-      uses: actions/setup-python@v5
+      uses: jmarrec/setup-python@v5
```
Comment (jmarrec) [inline .github/workflows/linux_release.yml:33]: That's the main change.

File: .github/workflows/linux_release.yml [hunk 5]
```diff
@@ -51,6 +51,10 @@ jobs:
         echo "Using Apt to install dependencies"
         sudo apt-get update
         sudo apt-get install texlive texlive-xetex texlive-science libxkbcommon-x11-0 xorg-dev libgl1-mesa-dev
+        if [[ "${{ matrix.os }}" == "ubuntu-24.04" ]]; then
+          # https://github.com/actions/runner-images/issues/10025
+          echo "FC=gfortran-13" >> $GITHUB_ENV
+        fi
```
Comment (jmarrec) [inline .github/workflows/linux_release.yml:57]: Workaround for Ubuntu24.04 not picking up fortran by default (filed upstream)

**Comment by jmarrec** [comment]
@Myoldmopar rebase onto develop, hopefully CI comes all green now that python 3.12.2 is there

File: CMakeLists.txt (hunk 1 above)
Comment (Myoldmopar) [reply CMakeLists.txt:194]: Not anymore!

File: scripts/package_tests/runner.py (hunk 2 above)
Comment (Myoldmopar) [reply scripts/package_tests/runner.py:91]: Nice.

**Review by Myoldmopar — APPROVED** [review]
Looks good.  Although Linux isn't responding to this PR yet.  Which is...odd.  I'll check back tomorrow.

**Comment by jmarrec** [comment]
Linux came around. The delay was astonishing though

**Comment by Myoldmopar** [comment]
Alright, all green.  Let's merge this and we're on to 3.12.2!  Now to figure out that silly `boto` issue...

**Pull request merged**